lib/jsdom/living/helpers/ce-reactions.js

```diff
@@ -0,0 +1,680 @@
+    "pathname",
+    "search",
+    "hash"
+  ],
+  HTMLHeadingElement: ["align"],
+  HTMLTextAreaElement: [
+
+    "autocomplete",
+    "autofocus",
+    "cols",
+    "dirName",
+    "disabled",
+    "inputMode",
+    "maxLength",
+    "minLength",
+    "name",
+    "placeholder",
+    "readOnly",
+    "required",
+    "rows",
+    "wrap",
+    "defaultValue",
+    "value"
+  ],
+  HTMLButtonElement: [
+
+    "autofocus",
+    "disabled",
+    "formAction",
+    "formEnctype",
+    "formMethod",
+    "formNoValidate",
+    "formTarget",
+    "name",
+    "type",
+    "value"
+  ],
+  HTMLUListElement: [
+    "compact",
+    "type"
+  ],
+  Node: [
+    "nodeValue",
+    "textContent",
+    "normalize",
+    "cloneNode",
+    "insertBefore",
+    "appendChild",
+    "replaceChild",
+    "removeChild"
+  ],
+  HTMLModElement: [
+    "cite",
+    "dateTime"
+  ],
+  HTMLScriptElement: [
+    "src",
+    "type",
+    "noModule",
+    "async",
+    "defer",
+    "crossOrigin",
+    "text",
+    "nonce",
+    "integrity",
+    "charset",
+    "event",
+    "htmlFor"
+  ],
+  HTMLPreElement: ["width"],
+  HTMLEmbedElement: [
+    "src",
+    "type",
+    "width",
+    "height",
+    "align",
+    "name"
+  ],
+  HTMLVideoElement: [
+    "width",
+    "height",
+    "poster",
+    "playsInline"
+  ],
+  HTMLFrameElement: [
+    "name",
+    "scrolling",
+    "src",
+    "frameBorder",
+    "longDesc",
+    "noResize",
+    "marginHeight",
+    "marginWidth"
+  ],
+  DOMTokenList: [
+    "add",
+    "remove",
+    "toggle",
+    "replace",
+    "value"
+  ],
+  HTMLSlotElement: ["name"],
+  Document: [
+    "createElement",
+    "createElementNS",
+    "importNode",
+    "adoptNode",
+    "title",
+    "dir",
+    "body",
+    "open",
+    "close",
+    "write",
+    "writeln",
+    "designMode",
+    "execCommand",
+    "fgColor",
+    "linkColor",
+    "vlinkColor",
+    "alinkColor",
+    "bgColor",
+
+    // Inherited from: ParentNode mixin
+    "prepend",
+    "append"
+  ],
+  DocumentType: [
+    // Inherited from: ChildNode mixin
+    "before",
+    "after",
+    "replaceWith",
+    "remove"
+  ],
+  DocumentFragment: [
+    // Inherited from: ParentNode mixin
+    "prepend",
+    "append"
+  ],
+  HTMLFieldSetElement: [
+    "disabled",
+    "name"
+  ],
+  HTMLOListElement: [
+    "reversed",
+    "start",
+    "type",
+    "compact"
+  ],
+  HTMLOptGroupElement: [
+    "disabled",
+    "label"
+  ],
+  HTMLOptionElement: [
+    "disabled",
+    "label",
+    "defaultSelected",
+    "value",
+    "text"
+  ],
+  NamedNodeMap: [
+    "setNamedItem",
+    "setNamedItemNS",
+    "removeNamedItem",
+    "removeNamedItemNS"
+  ],
+  CustomElementRegistry: [
+    "define",
+    "upgrade"
+  ],
+  Attr: [
+    "value",
+
+    // Inherited from: Node
+    "nodeValue",
+    "textContent"
+  ],
+  CSSStyleDeclaration: [
+    "cssText",
+    "setProperty",
+    "removeProperty",
+    "cssFloat"
+  ]
+};
+
+function wrapWithCEReactions(original) {
```

Comment:
Since webidl2js currently doesn't process the `[CEReactions]` annotation, for the purpose of this PR I used 2 different approaches:

**Interface patching:** Before exporting the interfaces, each property that requires `[CEReactions]` is patched. The approach requires to keep a mapping of interfaces and properties to patch which is not ideal. This approach also doesn't work for interfaces using named and index setters (eg. `DOMStringMap`), because there is not known properties to patch in advance on the interface.

**Impl patching:** Instead of patching the generated interface, in this approach the Impl class prototype is directly patched. It allows solving the issue described above for named and index setters: https://github.com/jsdom/jsdom/pull/2548/files#diff-2be81b507625c47c1dab5327ca4ace8aR65. If we were to use this approach everywhere it will require each impl to have list of all the properties that are required to be patched (vs. a central place).

A final approach would be to delegate this webidl2js, but I didn't spend enough time thinking about how to implement this.

Thoughts?

Reply:
I would implement this in webidl2js, similar to [Reflect], with the following design: it outputs a method body of roughly:

```js
// do all argument conversions etc.

impl._ceReactionsPreSteps();

try {
  ... original stuff, including calling impl[methodName](...args);
} finally {
  impl._ceReactionsPostSteps();
}
```

Then, in jsdom, we would implement `_ceReactionsPreSteps()` and `_ceReactionsPostSteps()` in Node-impl.js.

Reply:
The better architecture would be to add plugins to webidl2js, to let jsdom write a plugin that did this CEReactions stuff. Maybe you would want to give that a try, but it sounds hard to me. The _ceReactionsPre/PostSteps insertion is my attempt at something simple but still slightly decoupled.

I guess another approach would be to add an option to webidl2js's generate, e.g.

```js
transformer.generate(outputDir, {
  processCEReactions: normalMethodBody => expandedMethodBody
});
```

That's pretty simple, and even more decoupled.

(A "plugin" architecture is probably like that, but more general, so that you can intercept any extended attribute, applied to any IDL construct, not just CEReactions applied to operations.)